kubernetes/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterScheduler.scala

```diff
@@ -0,0 +1,236 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.cluster.kubernetes
+
+import java.io.File
+import java.util.Date
+import java.util.concurrent.atomic.AtomicLong
+
+import io.fabric8.kubernetes.client.{ConfigBuilder, DefaultKubernetesClient, KubernetesClient}
+import io.fabric8.kubernetes.api.model.{PodBuilder, ServiceBuilder}
+import io.fabric8.kubernetes.client.dsl.LogWatch
+import org.apache.spark.deploy.Command
+import org.apache.spark.deploy.kubernetes.ClientArguments
+import org.apache.spark.{io, _}
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+
+import collection.JavaConverters._
+import org.apache.spark.util.Utils
+
+import scala.util.Random
+
+private[spark] object KubernetesClusterScheduler {
+  def defaultNameSpace = "default"
+  def defaultServiceAccountName = "default"
+}
+
+/**
+  * This is a simple extension to ClusterScheduler
+  * */
+private[spark] class KubernetesClusterScheduler(conf: SparkConf)
+    extends Logging {
+  private val DEFAULT_SUPERVISE = false
+  private val DEFAULT_MEMORY = Utils.DEFAULT_DRIVER_MEM_MB // mb
+  private val DEFAULT_CORES = 1.0
+
+  logInfo("Created KubernetesClusterScheduler instance")
+
+  var client = setupKubernetesClient()
+  val driverName = s"spark-driver-${Random.alphanumeric take 5 mkString("")}".toLowerCase()
+  val svcName = s"spark-svc-${Random.alphanumeric take 5 mkString("")}".toLowerCase()
+  val nameSpace = conf.get(
+    "spark.kubernetes.namespace",
+    KubernetesClusterScheduler.defaultNameSpace)
+  val serviceAccountName = conf.get(
+    "spark.kubernetes.serviceAccountName",
+    KubernetesClusterScheduler.defaultServiceAccountName)
+
+  // Anything that should either not be passed to driver config in the cluster, or
+  // that is going to be explicitly managed as command argument to the driver pod
+  val confBlackList = scala.collection.Set(
+    "spark.master",
+    "spark.app.name",
+    "spark.submit.deployMode",
+    "spark.executor.jar",
+    "spark.dynamicAllocation.enabled",
+    "spark.shuffle.service.enabled")
+
+  def start(args: ClientArguments): Unit = {
+    startDriver(client, args)
+  }
+
+  def stop(): Unit = {
+    client.pods().inNamespace(nameSpace).withName(driverName).delete()
```

Comment:
I think if we do this here in cluster mode we will shut ourselves down and risk not actually finishing the `stop()` method. It's also possible that the `SparkContext` object and its associated components like the scheduler can be stopped, but the JVM still is doing other things afterwards. Therefore I don't think we should be deleting the driver pod here.